docs/doc_index.py

```diff
@@ -0,0 +1,163 @@
+    """Main entry point for index.py file that replaces a standard index.rst file.
+
+    The contents of this doc will be used to generate corresponding index.rst
+
+    Parameters
+    ----------
+    title : str or tuple[str, str]
+        Either a title used within `..title::` directive or a tuple of (title, underline_char).
+        In the second case, the underline_char will be used to do the sphinx header by placing
+        it len(title) times under the title.
+    options : str or list[str]
+        List of options like `:maxdepth:` for the toctree.
+    entries : list[str or example_entry(...)]
+        Toctree of subpages, can be either represented by regular strings or by
+        `example_entry()` that allows to put the reference from operator to given notebook.
+
+        Entries come in three form:
+          * a path to index file (without an extension at the end), for example: "operations/index"
+            will expect to lead into `operations/index.py` file and process it recursively
+          * a path with extension to either .rst or .ipynb file - they will be inserted as is
+          * an example_entry()
+
+    """
+    global doc_return_value
+    doc_return_value = Doc(title, options, entries)
+
+
+def example_entry(name, operator_ref = None):
+    """Place given notebook or doc page in the toctree and optionally add a reference from operator
+    documentation to that notebook or page.
+
+    Parameters
+    ----------
+    name : str
+        Name of jupyter notebook or rst file, must contain proper extension.
+    operator_ref : OpReference or List[OpReference], optional
+        Optional reference, defined by `op_reference()` call, by default None
+    """
+    return ExampleEntry(name, operator_ref)
+
+
+def op_reference(operator, docstring):
+    """Add a reference from operator to this notebook with specified docstring.
+
+    Parameters
+    ----------
+    operator : str
+        Name of operator without nvidia.dali prefix, for example fn.resize or fn.gaussian_blur
+    docstring : str
+        Text that would appear in the see also block for given link.
+    """
+    return OpReference(operator, docstring)
+
+
+def _obtain_doc(py_file):
+    """Extract the doc() definition from index.py file"""
+    with open(py_file, 'r') as f:
+        doc_file = f.read()
+        exec(doc_file)
+        return doc_return_value
+
+
+def _document_examples(path, result_dict={}):
+    py_file = path + ".py"
+    rst_file = path + ".rst"
+
+    doc_contents = _obtain_doc(py_file)
+    tab = " " * 3
+    with open(rst_file, "w") as f:
+        if isinstance(doc_contents.title, str):
+            f.write(f".. title:: {doc_contents.title}\n\n")
+        else:
+            title, level = doc_contents.title
+            f.write(f"{title}\n{level * len(title)}\n\n")
+
+        f.write(f".. toctree::\n")
+        if not isinstance(doc_contents.options, list):
+            doc_contents.options = [doc_contents.options]
```

Comment:
I would move it to `_obtain_doc` so the result doesn't need any post processing here.